dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.SkillDefinition;
+
+namespace Microsoft.SemanticKernel.Orchestration;
+
+/// <summary>
+/// Standard Semantic Kernel callable plan.
+/// Plan is used to create trees of <see cref="ISKFunction"/>s.
+/// </summary>
+public sealed class Plan : ISKFunction
+{
+    /// <summary>
+    /// State of the plan
+    /// </summary>
+    [JsonPropertyName("state")]
+    public ContextVariables State { get; } = new();
+
+    /// <summary>
+    /// Steps of the plan
+    /// </summary>
+    [JsonPropertyName("steps")]
+    internal List<Plan> Steps { get; } = new();
+
+    /// <summary>
+    /// Named parameters for the function
+    /// </summary>
+    [JsonPropertyName("named_parameters")]
+    public ContextVariables NamedParameters { get; set; } = new();
+
+    public bool HasNextStep => this.NextStep < this.Steps.Count;
+
+    public int NextStep { get; private set; } = 0;
+
+    #region ISKFunction implementation
+
+    /// <inheritdoc/>
+    [JsonPropertyName("name")]
+    public string Name { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("skill_name")]
+    public string SkillName { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("description")]
+    public string Description { get; set; } = string.Empty;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("is_semantic")]
+    public bool IsSemantic { get; internal set; } = false;
+
+    /// <inheritdoc/>
+    [JsonPropertyName("request_settings")]
+    public CompleteRequestSettings RequestSettings { get; internal set; } = new();
+
+    #endregion ISKFunction implementation
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a goal description.
+    /// </summary>
+    /// <param name="goal">The goal of the plan used as description.</param>
+    public Plan(string goal)
+    {
+        this.Description = goal;
+        this.SkillName = this.GetType().FullName;
+        this.Name = goal;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Plan"/> class with a goal description and steps.
+    /// </summary>
+    /// <param name="goal">The goal of the plan used as description.</param>
+    /// <param name="steps">The steps to add.</param>
+    public Plan(string goal, params ISKFunction[] steps) : this(goal)
+    {
+        this.AddStep(steps);
+    }
+
+    /// <summary>
```

Comment:
Remove the method or use the ISKFunction impl.

Reply:
@RogerBarreto which method are you suggesting we remove? The third constructor?
` public Plan(string goal, params Plan[] steps)`

Reply:
Ah, I gotcha. Plan[] is a ISKFunction

Reply:
Passing around ISKFunction causes the Plan properties to be lost. I'll be adding these back in.